Desktop: preview mode fix

Runo-saduwa opened the pull request:
**DESCRIPTION:**
This patch is a fix for the issue #2467 . Previously before this solution, the Edit menu items were not disabled on PREVIEW, this caused users to be able to use short cuts such as CTRL+B, CTRL+` on PREVIEW mode which is an unwanted behavior.

**HOW TO TEST**
Use the following test cases to confirm this solution::

- Add a note to a notebook and click the layout toggler to PREVIEW:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**
- Add a note to a notebook and click the layout toggler to EDITOR:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **enabled**
- Add a note to a notebook and click the layout toggler to EDITOR AND PREVIEW MODE:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **enabled**
- Imported a HTML note  and click the layout toggler to EDITOR/PREVIEW MODE/SPLIT:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**
- CREATE an Empty Notebook (book without notes)
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**

mic704b commented on ElectronClient/app.js:
```diff
@@ -1178,7 +1182,16 @@ class Application extends BaseApplication {
+					menuItem.enabled = false;
+				} else {
+					menuItem.enabled = true;
+				}
```
This can be one line  `menuItem.enabled = (viewer === 'viewer');

mic704b commented on ElectronClient/app.js:
```diff
@@ -1178,7 +1182,16 @@ class Application extends BaseApplication {
+					menuItem.enabled = true;
+				}
+			} else {
+
```
unnecessary white space

mic704b commented on ElectronClient/app.js:
```diff
@@ -281,6 +281,9 @@ class Application extends BaseApplication {
 			Setting.setValue('noteVisiblePanes', newState.noteVisiblePanes);
+			const viewer = newState.noteVisiblePanes[0];
+			this.updateMenuItemStates(viewer);
+
```
unnecessary white space

mic704b commented:
@Runo-saduwa nice work, didn't check the logic, just a few style comments.

Runo-saduwa commented:
I am still trying to set up Web Clipper to test this. as soon as i successfully set up Web Clipper on my browser, this issue will be resolved

tessus commented:
Just use the "normal" web clipper with your production Joplin insgtance. Then create a HTML note as I explained earlier, export it (jex), and import it into your dev Joplin instance.

tessus commented:
Or use this one:

[html-test.zip](https://github.com/laurent22/joplin/files/4253313/html-test.zip)

Runo-saduwa commented:
Thank You, I will give updates shortly

Runo-saduwa commented:
> Just use the "normal" web clipper with your production Joplin insgtance. Then create a HTML note as I explained earlier, export it (jex), and import it into your dev Joplin instance.

Special thanks to @tessus

Runo-saduwa commented:
@PackElend Please label me, thanks.

Runo-saduwa commented:
Thanks for the quick review @tessus

laurent22 commented on ElectronClient/app.js:
```diff
@@ -1166,21 +1169,23 @@ class Application extends BaseApplication {
 		const selectedNoteIds = state.selectedNoteIds;
 		const note = selectedNoteIds.length === 1 ? await Note.load(selectedNoteIds[0]) : null;
-
 		for (const itemId of ['copy', 'paste', 'cut', 'selectAll', 'bold', 'italic', 'link', 'code', 'insertDateTime', 'commandStartExternalEditing', 'showLocalSearch']) {
```
Please remove empty line from commit.

laurent22 commented on ElectronClient/app.js:
```diff
@@ -1166,21 +1169,23 @@ class Application extends BaseApplication {
+			}
 		}
-
 		const menuItem = Menu.getApplicationMenu().getMenuItemById('help:toggleDevTools');
```
Please remove empty line from commit.

laurent22 commented on ElectronClient/app.js:
```diff
@@ -1166,21 +1169,23 @@ class Application extends BaseApplication {
+				menuItem.enabled = false;
+			} else if (noteCheck === true || viewer !== 'viewer') {
+				menuItem.enabled = true;
+			}
```
Sorry I can't wrap my head around this logic. There's an implied "else" case - in which case is it executed? Never? Then why use an else/if?

Also please could you clarify what "noteCheck" means? Is it possible to find a name that better expresses your intent?

Runo-saduwa replied:
about the else/if statement, i have changed that to else and the logic still works fcine. noting was broken.

and the `noteCheck` variable is used to store the boolean value that will be returned from this statement `!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;`, which is used to check if there is a note and also if the note is a HTML note.

i have added a comment to explain that part.

mic704b replied:
Hi @Runo-saduwa, I don't want to interfere, but just to suggest maybe something like 'isHtmlNote' might be a better name for 'noteCheck'.  Or ideally maybe don't bother with making a temporary variable at all because you don't need it again.

Another style tip, consider using just 'if (x)' instead of if (x === true)' unless you have an explicit reason for doing the latter. Writing it this way will also help you think of a good name for it.

Runo-saduwa replied:
@mic704b that’s a great suggestion, thanks. I’ll with your suggestions. 

isHtmlNote is cool

Runo-saduwa replied:
The code has been improved again, this time I improved the variable name to something meaningful (isHtmlNote) and I also replaced  the If statement with a sweet ternary operator. 

You can go ahead and test it, noting is broken, everything is in order.

mic704b replied:
>  I also replaced the If statement with a sweet ternary operator

Better, but switch it around and it would be even better.

mic704b replied:
And remove the comment about `noteCheck`

Runo-saduwa replied:
Yeah Thanks, I just did that

laurent22 commented on ElectronClient/app.js:
```diff
@@ -1178,7 +1181,8 @@ class Application extends BaseApplication {
 			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			!isHtmlNote || viewer === 'viewer' ?  menuItem.enabled = false : menuItem.enabled = true;
```
This is like writing `true ? x = false : x = true;` or you know, you could just write `x = false`. This is the same here, you could simply write `menuItem.enabled = !isHtmlNote && viewer !== 'viewer'`.

Runo-saduwa replied:
I just tried it, it doesn’t work as expected, it breaks the expected functionality. I think this is the simplest and most readable way to solve the issue and maintain its expectated behavior.

mic704b replied:
`menuItem.enabled = isHtmlNote && viewer !== 'viewer';`

laurent22 replied:
ah, now I see what the problem is. We don't want the items to be enabled if it is an HTML note or if it the viewer. The problem here is that `isHtmlNote` is false when it's an HTML note (and true when it's a Markdown note), so Runo you have it backward.

Essentially you wrote this: `isHtmlNote = note.type === 'markdown'` - doesn't that tell you that it's in fact a Markdown note, and not an HTML one? So full condition should be `const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;`

laurent22 replied:
And I insist on getting rid of `true ? x = false : x = true;`

mic704b replied:
Perfect!

> it doesn’t work as expected, it breaks the expected functionality

@Runa-saduwa whenever you think this, it means it's time to look closer to find out why!

Runo-saduwa replied:
> And I insist on getting rid of true ? x = false : x = true;

i apologize, I have figured it out. I’ve gotten rid this

Runo-saduwa replied:
@mic704b you’re very correct. Thank you for that, I was a bit hasty in trying to solve the issue that brought about the PR

mic704b replied:
All good, now make sure to read Laurent's comment above closely, and adjust the code as described if it makes sense.

Laurent said:
> Essentially you wrote this: `isHtmlNote = note.type === 'markdown'` - doesn't that tell you that it's in fact a Markdown note, and not an HTML one? So full condition should be `const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;`

Runo-saduwa replied:
I’ve made the adjustments.

If this isn’t okay, I still have another way to do this, I’ve looked closely at the code and I’ve seen different ways to achieve what we want.

laurent22 commented on ElectronClient/app.js:
```diff
@@ -1178,7 +1181,8 @@ class Application extends BaseApplication {
 			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const isHtmlNote = !!note && note.markup_language !== MarkupToHtml.MARKUP_LANGUAGE_HTML;
+			menuItem.enabled = isHtmlNote && viewer !== 'viewer';
```
I was expecting I could merge this, but after all this time it's still not good. I don't know what to say since we've already went over this. If something's not clear about what should happen please ask.

Also please rename the "viewer" variable, it doesn't make sense. It's like having a variable called "plane" and setting it to "car".

Runo-saduwa replied:
My bad, I thought I understood your last comment well enough, that was why I didn’t want to bother you.

But based on what you said, I could change the `isHtmlNote` to `isMarkDownNote` And turn the final statement to:

```isMarkDownNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN```

With the above code,  when the note is a HTML note, `isMarkDownNote` will be `false`And when the note is a Markdown note, `isMarkDownNote` will be `true`.

How do you see this approach, is it okay?

Runo-saduwa replied:
I’ll change that to `editorMode`

tessus replied:
> I’ll change that to editorMode

This is also wrong. Dude, you have 3 layouts: viewer, split, editor. So how would you call the mode? I'd say `layout` or `layoutMode`.

tessus replied:
> How do you see this approach, is it okay?

No, it's not. You don't have to rename the variable `isHtmlNote`. Anyway, I'm too tired now. Ttyl.

mic704b replied:
Hi @Runo-saduwa 

tessus is right about the layouts and I think you've sorted that out now.

Concerning your question about `isHtmlNote`, tessus is also right, the variable you have in the code above is correctly named.  

However, as you suggest, you could instead implement a `isMarkdownNote`.  Either variation will work. Which one you choose to go with depends on how you use it thereafter, which one makes the your following code simpler to read and understand.

tessus replied:
Hi @mic704b . Runo has already made all the requested changes. The code above is outdated. I think the best way is to just click on `Files changed` and start a new review.

It's a bit confusing that these conversations are not automarically resolved as soon as the code does not exist anymore.

mic704b replied:
Thanks @tessus, I'm a bit behind.  I just wanted to respond to Runo's question.  Although it wasn't resolving the issues that were being discussed on this thread, the use of `isMarkdownNote` was an OK option to consider for other reasons.

PackElend commented:
@Runo-saduwa can you summarize you changes?
Please also include the comments / reasons.
This PR has a lot of comments, so it is quite challenging to follow it.
thank you

Runo-saduwa commented:
>  So full condition should be const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;
based on this comment by Laurent, i have changed the code to use this.
``` 
const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML
```
i also renamed the `viewer` variable to `layout` because it connotes the meaning of the value it stores.

Then i modified the logic for enabling and disabling the `menuItem` with this condition 
```!isHtmlNote && layout !== 'viewer' && !!note```
which causes the same side-effects as this:
```
 if(isHtmlNote || layout === 'viewer' || !note){ 
        menuItem.enabled = false;
 } else {
	menuItem.enabled = true;
 }
```

After doing all these changes, i tested with the following cases and confirmed that it works properly:

- Add a note to a notebook and click the layout toggler to PREVIEW:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**
- Add a note to a notebook and click the layout toggler to EDITOR:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **enabled**
- Add a note to a notebook and click the layout toggler to EDITOR AND PREVIEW MODE:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **enabled**
- Imported a HTML note  and click the layout toggler to EDITOR/PREVIEW MODE/SPLIT:
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**
- CREATE an Empty Notebook (book without notes)
EXPECTED RESULT: 'copy', 'paste', 'cut', 'select all', 'bold', 'italic', 'link', 'code', 'insert Date Time', 'Start External Editing', 'show Local Search' should be **disabled**

Runo-saduwa commented on ElectronClient/app.js:
```diff
@@ -1178,7 +1181,8 @@ class Application extends BaseApplication {
 			const menuItem = Menu.getApplicationMenu().getMenuItemById(`edit:${itemId}`);
 			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;
```
>  So full condition should be const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;
based on this comment by Laurent, i have changed the code to use this.
``` 
const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML
```

Runo-saduwa commented on ElectronClient/app.js:
```diff
@@ -1166,7 +1169,7 @@ class Application extends BaseApplication {
 	}
 
-	async updateMenuItemStates(state = null) {
+	async updateMenuItemStates(layout, state = null) {
```
i also renamed the `viewer` variable to `layout` because it connotes the meaning of the value it stores.

Runo-saduwa commented on ElectronClient/app.js:
```diff
@@ -1178,7 +1181,8 @@ class Application extends BaseApplication {
 			if (!menuItem) continue;
-			menuItem.enabled = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_MARKDOWN;
+			const isHtmlNote = !!note && note.markup_language === MarkupToHtml.MARKUP_LANGUAGE_HTML;
+			menuItem.enabled = !isHtmlNote && layout !== 'viewer' && !!note;
```
Then i modified the logic for enabling and disabling the `menuItem` with this condition 
```!isHtmlNote && layout !== 'viewer' && !!note```
which causes the same side-effects as this:
```
 if(isHtmlNote || layout === 'viewer' || !note){ 
        menuItem.enabled = false;
 } else {
	menuItem.enabled = true;
 }
```

tessus commented:
Anything left to do here or can we merge this?

laurent22 commented:
We got there eventually! Thanks @Runo-saduwa for the fix and everyone for reviewing.
